### DISCUSSION TITLE
Congratulations and future

### DISCUSSION THREAD

**Discussion by frederikhors** [jycouet/kitql#36] [General]
I discovered this project a few hours ago. I saw your video on YouTube and I must say that it all looks wonderful to me.

I come from urql and before that apollo. I am fine with urql but the team that created it works mainly with React and the implementation for Svelte is very verbose and sometimes not as functional as one imagines.

Your project got me excited. I am not joking or exaggerating this statement.

And it's already great for professional use.

If you don't mind I have prepared a list of features that I think are urgent for the most demanding projects and I hope to be able to contribute too.

I would only like one certainty in return from you: that you do not abandon the project after having worked on it for a while: it is too good to be abandoned!!

Backlog (we can decide to make them issues each of these and to use the project tab if you want to better prioritize each of them):

1. use [docusaurus](https://github.com/facebook/docusaurus) for a very good docs (see [pnpm docs](https://pnpm.io/motivation) to understand what I mean) or maybe we can use SvelteKit with some template

1. introduce a `cache-and-network` policy (better explained [here](https://formidable.com/open-source/urql/docs/basics/document-caching/#request-policies))

1. persist stores to localstorage or indexedDB (to have data immediately as soon as I return to the app in the future (even if it is stale data))

1. [invalidation](https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#invalidating-entities) or [this one](https://react-query.tanstack.com/guides/query-invalidation)

1. [paginated](https://react-query.tanstack.com/guides/paginated-queries) and [infinite queries](https://react-query.tanstack.com/guides/infinite-queries) the same as [URQL](https://formidable.com/open-source/urql/docs/graphcache/local-resolvers/#pagination)

1. [placeholder query data](https://react-query.tanstack.com/guides/placeholder-query-data) the same URQL obtains with [Schema Awareness](https://formidable.com/open-source/urql/docs/graphcache/schema-awareness/)

1. [background fetching indicator](https://react-query.tanstack.com/guides/background-fetching-indicators)

1. nice-to-have: [APQ](https://formidable.com/open-source/urql/docs/advanced/persistence-and-uploads/#automatic-persisted-queries) 

1. [offline support](https://formidable.com/open-source/urql/docs/graphcache/offline/)

1. more coming from my head soon...

If something is difficult to do for the SSR part I would start doing it for the runtime part only for now.

What do you think?

**Comment by jycouet** [comment]
Thx a lot @frederikhors for opening an issue like this. It's already a nice contribution 👍


I'm also coming from apollo then urql, so I know a lot about your background. We have the same !

Maybe, this issue could go to [discussions](https://github.com/jycouet/kitql/discussions) and we will open issues one by one when we agree on each point? (I'll try to arrange later)

I'm spending quite some time on this `KitQL` and use it on multiple projects... I don't want to abandon it! I even update a thread there: https://twitter.com/jycouet. 👍. 
The best way for an OSS library to not be abandoned is usage and contribution by the community. 
@all-contributors please add @frederikhors for ideas

About the your **Backlog**, even if I know most of the topics, I need to dig a bit more on some.
Let me come back later in discussions > issues

Thx a lot ✅

[a comment by an automated account is not shown]

**Comment by frederikhors** [comment]
You sure are a very nice person.

I agree to move this issue to discussions (indeed forgive me if I opened it: I hadn't even noticed that there was a discussions tab). Indeed, if you give me permission to move it in discussions, I can do it myself.

I agree 100% of what you said.

Just one thing: can we decide together the most urgent things to do here https://github.com/jycouet/kitql/projects/1?

**Reply by jycouet** [reply]
I have to say that it's a bit a mess... Issues / Discussions / Projects... What is the best to use? When? (I'm also discovering...)
So I added the stuff I know already that needs to be there asap:
- https://github.com/jycouet/kitql/discussions/20
- https://github.com/users/jycouet/projects/2
- https://github.com/jycouet/kitql/issues?q=is%3Aopen+is%3Aissue+label%3Aenhancement

For all you remaining points, I will do a dedicated thread here to speak a bit about it before adding them 🚀

(I see the potential of `project`, but maybe it's a bit early to use it ^^)

**Reply by frederikhors** [reply]
I cannot wait!

**Reply by jycouet** [reply]
Your time to comment :)

**Comment by jycouet** [comment]
2/ introduce a cache-and-network policy (better explained here)

Today `KitQL` already has:
https://github.com/jycouet/kitql/blob/b7821c4e522fb0e987308bb9f24a0db8b4434940/examples/demo1/src/lib/components/Continents.svelte#L12-L14

And actually I should always be in `cache-and-network`.
Do you need all others?

Maybe one need is `no-caching`?

**Reply by jycouet** [reply]
In the end, I did it. It's in `"@kitql/all-in": "0.1.9"` 🚀🚀🚀

**Reply by frederikhors** [reply]
I think we need `cache-only` (with no TTL) and `network-only`.

**Reply by jycouet** [reply]
It's in ;)

**Comment by jycouet** [comment]
4/ invalidation

The demo is not up to date, but `KitQL` has
https://github.com/jycouet/kitql/blob/b7821c4e522fb0e987308bb9f24a0db8b4434940/packages/client/src/lib/components/Continents.svelte#L13-L15

I think that it's sufficient. Maybe I'm missing something?

(I need to update the demo)

**Reply by frederikhors** [reply]
I think that may be enough, but I have to try it (so far I haven't tried anything). For this I would like to create (together with you) a very complex application to find all the useful missing features.

**Reply by jycouet** [reply]
Yep, & with a doc it will be easier for everyone to understand what exist, what is in mind and what is missing.

**Comment by jycouet** [comment]
5/ paginated and infinite queries the same as URQL

I need to dig more in the topic. Today, I have my own solution (in projects), I don't see what is really needed in KitQL.

**Reply by frederikhors** [reply]
In the demo app we can try this solution (if you want of course).

**Comment by jycouet** [comment]
6/ placeholder query data the same URQL obtains with Schema Awareness

This is for `Partial Results` right?
How "critical" is this? I would delay this one.

Any comments? Urgency?

**Reply by frederikhors** [reply]
This is one of the biggest points in favor of solutions like urql or apollo for me. And that's one of the reasons I really enjoy working with Graphcache (normalized data) and schema awareness. Although it makes everything much heavier (in terms of development and runtime).

The UX show in the hands of the user is that the application looks native and almost offline (if the internet is fast enough). The loading spinners disappear completely and everything seems immediate.

Ultimately that's why I chose to complicate my life by choosing to develop single page apps (SPAs) instead of multi-page (server rendered) apps (MPAs).

A very simple example is the list > detail. As soon as I enter the details I can show the data that I have already shown in the list and download in the background (almost without the person noticing, especially if I do it in the prefetch) the data that did not appear in the list.

I know that you understand what I mean and I also know that it is difficult to understand well how to organize this thing. But that would be great.

**Reply by jycouet** [reply]
Yes, I know what you mean and I would love to have something like this.
But maybe with another approach! Simpler :) 

(I already have something in mind, and the more I think about it the more I think it's not that hard to have ^^)

Side note: For me SPAs or MPAs... Now it's SvelteKit handling both seamless. 🚀

**Reply by frederikhors** [reply]
I agree with you 100%. I think you are another me but much better than me! 😄

**Comment by jycouet** [comment]
8/ Persisted Queries

Your link is an `Apollo` custom way of doing right?
I was planning to go more in that direction: https://github.com/valu-digital/graphql-codegen-persisted-query-ids

Any comments? Urgency?

**Reply by frederikhors** [reply]
It looks perfect. No comment. Not urgent.

**Comment by jycouet** [comment]
9/ Offline support

Yes, I would like to have something arround that. But when? I don't know yet!

Any comments? Urgency?

**Reply by frederikhors** [reply]
Not urgent at all.
